### ISSUE TITLE
Type confusion when initializing a variable with && and comparison

### ISSUE THREAD

**Issue by phistuck** [google/closure-compiler#1213]
The following code -
```
var /** @type {boolean} */ typing = (10 && ((10 - (new Date(10)).getTime()) < 7000));
```
Produces the following incorrect warning -
```
JSC_TYPE_MISMATCH: initializing variable
found   : (boolean|number)
required: boolean at line 1 character 37
var /** @type {boolean} */ typing = (10 && ((10 - (new Date(20)).getTime()) <...
                                     ^
```

However, the result of this expression will always be a `boolean`.

(This code is obviously nonsensical as it is now, but in practice, 10 and 20 are numeric variables.)

**Comment by dimvar** [comment]
Simplified example:

    var /** @type {boolean} */ typing = (10 && (1 < 2));

To fix this, the type checker would have to know that the left operand is a number literal that is not 0. That is trivial to check, but not generally useful. Finding that for an arbitrary left operand is more generally useful, but impossible to check statically.

**Comment by phistuck** [comment]
Oh, now I see that it will not always be a `boolean`. `var a = 0 && true` will be `0`. Huh. I guess I never noticed that.
Sorry!
